### PR TITLE
8330590: TextInputControl: previous word fails with Bhojpuri characters

### PR DESCRIPTION
**Pull request by andy-goryachev-oracle**
This change replaces Character.isLetterOrDigit(char) which fails with surrogate characters with  Character.isLetterOrDigit(int).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8330590](https://bugs.openjdk.org/browse/JDK-8330590): TextInputControl: previous word fails with Bhojpuri characters (**Bug** - P4)

### Reviewers
 * [Karthik P K](https://openjdk.org/census#kpk) (@karthikpandelu - Committer)
 * [Ambarish Rapte](https://openjdk.org/census#arapte) (@arapte - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1444/head:pull/1444` \
`$ git checkout pull/1444`

Update a local copy of the PR: \
`$ git checkout pull/1444` \
`$ git pull https://git.openjdk.org/jfx.git pull/1444/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1444`

View PR using the GUI difftool: \
`$ git pr show -t 1444`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1444.diff">https://git.openjdk.org/jfx/pull/1444.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1444#issuecomment-2067191578)

### CODE REVIEW

**Comment by kevinrushforth** [comment]
@karthikpandelu Can you review this? We'll also need a review by a "R"eviewer.

**Comment by andy-goryachev-oracle** [comment]
> Is this expected?

I think it might be a bug - even though it's unclear how many words the text "𑂦𑂷𑂔𑂣𑂳𑂩𑂲" contains, I would not expect it to go to the beginning of that segment.

I suspect the code in `TextInputControl.endOfNextWord(boolean)` is incorrect, and it needs a deeper re-write than the naive replacement with `isLetterOrDigit()`.

**Comment by andy-goryachev-oracle** [comment]
I think we need to fix endOf/nextWord as well, as the logic seems to be breaking with the surrogate pairs:

Screenshot 2024-04-29 at 09 48 33


The issue can also be seen with Awadhi: अवधी/औधी

**Comment by andy-goryachev-oracle** [comment]
Looking at the "next word" functionality across different applications on different platforms, it appears to be a wide variety of behaviors.

One vendor appears to be quite consistent - Microsoft.  Its word, word pad, notepad work exactly the same, with Word working the same across macOS and Win11.

JavaFX TextArea is inconsistent (by design) between macOS and Win11, but also is inconsistent with Swing's JTextArea.

If I were to fix the behavior (if we decide to fix the behavior of the nextWord function, that is), I would make it consistent with MS Word, but let's discuss.

For reference, here is the result of my testing.  Initially, the caret is placed at index 0 and the numbers in parentheses denote successive caret positions after ctrl-RIGHT (option-RIGHT) key presses.  An underline denotes a space, and a (nl) denotes a newline.

```
source
_english_english_eng:_end,_eng:_(nl)
(nl)
_eng


BreakIterator.getWordInstance()
_(1)english(2)_(3)english(4)_(5)eng(6):(7)_(8)end(9),(10)_(11)eng(12):(13)_(14)(nl)
(15)(nl)
(16)_(17)eng


text area (mac)
_english(1)_english(2)_eng(3):(4)_end(5),(6)_eng(7):(8)_(nl)
(9)(nl)
(10)_eng(11)


ms word (mac) 16.84 24041420 consistent with win11
_(1)english_(2)english_(3)eng(4):_(5)end(6),_(7)eng(8):_(9)(nl)
(10)(nl)
(11)_(12)eng(13)


text edit (mac)
_english(1)_english(2)_eng(3):_end(4),_eng(5):_(nl)
(nl)
(nl)_eng(6)


chrome (mac) <div contenteditable=true>
&nbsp;english(1)_english(2)_eng(3):(4)_end(5),(6)_eng(7):(8)_<br>
(9)<br>
_(10)eng(11)


eclipse (mac)
_(1)english_(2)english_(3)eng(4):_(5)end(6),_(7)eng(8):_(9)(nl)
(10)(nl)
(11)_(12)eng


JTextArea (mac)
_(1)english_(2)english_(3)eng(4):_(5)end(6),_(7)eng(8):_(9)(nl)
(nl)
_(10)eng


ms word 365 ver 2302 build 16.0.16130.20942 (win 11)
same as notepad (win 11)
same as wordpad (win 11)
_(1)english_(2)english_(3)eng(4):_(5)end(6),_(7)eng(8):_(9)(nl)
(10)(nl)
(11)_(12)eng


TextArea (win11)
_(1)english_(2)english_(3)eng(4):_(5)end(6),_(7)eng(8):_(9)(nl)
(10)(nl)
_(11)eng
```

**Comment by andy-goryachev-oracle** [comment]
@aghaisas would you please take a look at this also?

**Comment by andy-goryachev-oracle** [comment]
I've created https://bugs.openjdk.org/browse/JDK-8331951 to deal with the "next word" function issues.

**Review by arapte — CHANGES_REQUESTED** [review]
Fix looks good, providing minor comments.

File: modules/javafx.controls/src/main/java/javafx/scene/control/TextInputControl.java [hunk 1]
```diff
@@ -1743,4 +1742,15 @@ public int getAnchor() {
         }
     }
 
+    private static boolean isLetterOrDigit(String text, int ix, int len) {
```
[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/TextInputControl.java]: fixed, thank you

File: modules/javafx.controls/src/main/java/javafx/scene/control/TextInputControl.java [hunk 2]
```diff
@@ -1743,4 +1742,15 @@ public int getAnchor() {
...
+        if (ix < 0) {
+            // should not happen
+            return false;
+        } else if (ix >= text.length()) {
+            return false;
+        }
```
[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/TextInputControl.java:1751]: I prefer to keep one statement per line; the checks are needed here.

**Review by arapte — APPROVED** [review]

**Review by karthikpandelu — APPROVED** [review]

**Pull request closed without merging**